### PR TITLE
ASoC: SOF: Intel: discard SoundWire configuration if HDaudio codec is…

### LINKED ISSUE #4962: [BUG] Lenovo Ideapad slim 3i: Invalid SoundWire configuration reported on HDaudio platform.
Hello. I'm have a Lenovo ideapad slim 3i. Don't work sound (not found sound card). System Arch Linux, kernel 6.8.8 arch, pkg sof-firmware installed

log dmesg - https://pastebin.com/WWudLnkF

fragment sof log dmesg
```
[    5.739159] sof-audio-pci-intel-tgl 0000:00:1f.3: DSP detected with PCI class/subclass/prog-if info 0x040380
[    5.739188] sof-audio-pci-intel-tgl 0000:00:1f.3: SoundWire enabled on CannonLake+ platform, using SOF driver
[    5.739202] sof-audio-pci-intel-tgl 0000:00:1f.3: enabling device (0000 -> 0002)
[    5.739392] sof-audio-pci-intel-tgl 0000:00:1f.3: DSP detected with PCI class/subclass/prog-if 0x040380
[    5.739463] sof-audio-pci-intel-tgl 0000:00:1f.3: bound 0000:00:02.0 (ops i915_audio_component_bind_ops [i915])
[    5.746087] sof-audio-pci-intel-tgl 0000:00:1f.3: use msi interrupt mode
[    5.780311] sof-audio-pci-intel-tgl 0000:00:1f.3: DMICs detected in NHLT tables: 0
[    5.780317] sof-audio-pci-intel-tgl 0000:00:1f.3: hda codecs found, mask 5
[    5.783361] sof-audio-pci-intel-tgl 0000:00:1f.3: Firmware paths/files for ipc type 0:
[    5.783365] sof-audio-pci-intel-tgl 0000:00:1f.3:  Firmware file:     intel/sof/sof-adl-n.ri
[    5.783366] sof-audio-pci-intel-tgl 0000:00:1f.3:  Topology file:     intel/sof-tplg/sof-adl-rt711.tplg
[    5.783722] sof-audio-pci-intel-tgl 0000:00:1f.3: Firmware info: version 2:2:0-57864
[    5.783724] sof-audio-pci-intel-tgl 0000:00:1f.3: Firmware: ABI 3:22:1 Kernel ABI 3:23:0
[    5.783728] sof-audio-pci-intel-tgl 0000:00:1f.3: unknown sof_ext_man header type 3 size 0x30
[    5.787045] iwlwifi 0000:00:14.3: base HW address: bc:03:58:85:98:46
[    5.881061] sof-audio-pci-intel-tgl 0000:00:1f.3: Firmware info: version 2:2:0-57864
[    5.881082] sof-audio-pci-intel-tgl 0000:00:1f.3: Firmware: ABI 3:22:1 Kernel ABI 3:23:0
[    5.883475] soundwire_intel soundwire_intel.link.0: Bus clash for control word
[    5.883483] soundwire_intel soundwire_intel.link.0: Bus clash for data word
[    5.883571] soundwire_intel soundwire_intel.link.0: Bus clash for control word
[    5.883650] soundwire_intel soundwire_intel.link.0: Bus clash for control word
[    5.883734] soundwire_intel soundwire_intel.link.0: Bus clash for control word
[    5.883844] soundwire_intel soundwire_intel.link.0: Bus clash for control word
[    5.883921] soundwire_intel soundwire_intel.link.0: Bus clash for control word
[    5.884036] soundwire_intel soundwire_intel.link.0: Bus clash for control word
[    5.884112] soundwire_intel soundwire_intel.link.0: Bus clash for control word
[    5.884191] soundwire_intel soundwire_intel.link.0: Bus clash for control word
[    5.884290] soundwire_intel soundwire_intel.link.0: Bus clash for control word
[    5.904724] sof-audio-pci-intel-tgl 0000:00:1f.3: Topology: ABI 3:22:1 Kernel ABI 3:23:0
[    5.904793] sof-audio-pci-intel-tgl 0000:00:1f.3: error: can't connect DAI HDA3.OUT stream iDisp4
[    5.904795] sof-audio-pci-intel-tgl 0000:00:1f.3: error: failed to add widget id 0 type 27 name : HDA3.OUT stream iDisp4
[    5.904801] sof_sdw sof_sdw: ASoC: failed to load widget HDA3.OUT
[    5.904803] sof_sdw sof_sdw: ASoC: topology: could not load header: -22
[    5.904813] sof-audio-pci-intel-tgl 0000:00:1f.3: error: tplg component load failed -22
[    5.904816] sof-audio-pci-intel-tgl 0000:00:1f.3: error: failed to load DSP topology -22
[    5.904817] sof-audio-pci-intel-tgl 0000:00:1f.3: ASoC: error at snd_soc_component_probe on 0000:00:1f.3: -22
[    5.904828] sof_sdw sof_sdw: ASoC: failed to instantiate card -22
[    5.904895] sof_sdw sof_sdw: error -EINVAL: snd_soc_register_card failed -22
[    5.904905] sof_sdw: probe of sof_sdw failed with error -22
```

I`m installed https://github.com/thesofproject/sof-bin, this doesn't work

### PR DESCRIPTION
**Pull request by plbossart**
… reported

The machine driver and topology selection starts with I2S, then SoundWire and last uses HDaudio as a fallback. That assumes that the ACPI information is correc,t but there are of course exceptions to the rule.

On a Lenovo platform, an external HDaudio codec is detected, but the ACPI tables expose TWO RT711 jack codecs. This patch skips the SoundWire selection in case an external HDaudio codec is detected - which only works with the additional assumption that no one will mix HDaudio and SoundWire.

Closes: https://github.com/thesofproject/linux/issues/4962

### CODE REVIEW

**Comment by plbossart** [comment]
better if I push the version that compiles...

File: sound/soc/sof/intel/hda.c [hunk 1]
```diff
@@ -1305,6 +1305,7 @@ struct snd_soc_acpi_mach *hda_machine_select(struct snd_sof_dev *sdev)
 	u32 interface_mask = hda_get_interface_mask(sdev);
 	struct snd_sof_pdata *sof_pdata = sdev->pdata;
 	const struct sof_dev_desc *desc = sof_pdata->desc;
+	struct hdac_bus *bus = sof_to_bus(sdev);
```
Comment (ranj063) [inline sound/soc/sof/intel/hda.c:1308]: Looks good @plbossart . Just a small typo in the commit message, correc,t

Comment (plbossart) [reply sound/soc/sof/intel/hda.c:1308]: updated, thanks for noticing this @ranj063

**Review by bardliao — APPROVED** [review]

File: sound/soc/sof/intel/hda.h [hunk 2]
```diff
@@ -454,6 +454,8 @@
 #define SSP_SET_SFRM_CONSUMER	BIT(24)
 #define SSP_SET_CBP_CFP		(SSP_SET_SCLK_CONSUMER | SSP_SET_SFRM_CONSUMER)
 
+#define HDA_EXT_ADDR		0
+#define HDA_EXT_CODEC(x) ((x) & BIT(HDA_EXT_ADDR))
```
Comment (ujfalusi) [inline sound/soc/sof/intel/hda.h:458]: I guess address 0 is always for external codecs and BIT(1) and BIT(3) is not used?

**Review by ujfalusi — APPROVED** [review]

File: sound/soc/sof/intel/hda.h (hunk 2 above)
Comment (kv2019i) [reply sound/soc/sof/intel/hda.h:458]: In theory, a dual HDA codec system would use addr1 (SDI1), but in practise we've never had these with SOF. In most systems, codec count is limited to 3 with address 2 hardcodec to PCH display codec. For purpose of this patch (and check), this is ok.

**Review by kv2019i — APPROVED** [review]
Given there's a real-life case and the fact the HDA check is a runtime check (if codec is found, it's a pretty strong indicator to prioritize that in case static ACPI information conflicts).

File: sound/soc/sof/intel/hda.h (hunk 2 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda.h:458]: Yeah, I thought about the SDI1 case but we have no way to test it, so I chose to limit the scope to BIT(0).

**Review by ranj063 — APPROVED** [review]

**Pull request merged**